src/beatree/leaf/store.rs

```diff
@@ -0,0 +1,222 @@
+        let page = Box::new(Page::zeroed());
+
+        let mut command = Some(IoCommand {
+            kind: IoKind::Read(self.store_file.as_raw_fd(), pn.0 as u64, page),
+            handle: self.io_handle_index,
+            user_data: 0,
+        });
+
+        while let Some(c) = command.take() {
+            match self.io_sender.try_send(c) {
+                Ok(()) => break,
+                Err(TrySendError::Disconnected(_)) => panic!("I/O leaf store worker dropped"),
+                Err(TrySendError::Full(c)) => {
+                    command = Some(c);
+                }
+            }
+        }
+
+        // wait for completion
+        let completion = self
+            .io_receiver
+            .recv()
+            .expect("I/O leaf store worker dropped");
+        assert!(completion.result.is_ok());
+        let page = completion.command.kind.unwrap_buf();
+
+        LeafPage { inner: page }
+    }
+
+    // create a LeafStoreTx able to append and release leaves from the LeafStore
+    pub fn start_tx<'a>(&'a mut self) -> LeafStoreTx {
+        LeafStoreTx {
+            bump: &mut self.bump,
+            max_bump: self.max_bump,
+            free_list: &mut self.free_list,
+            to_allocate: vec![],
+            released: vec![],
+            exceeded: vec![],
+        }
+    }
+
+    // updates the mex supported page number based on the leaf store file size
+    pub fn update_size(&mut self) {
+        let file_size = self
+            .store_file
+            .metadata()
+            .expect("Error extracting metadata from LeafStore file")
+            .size() as usize;
+
+        self.max_bump = PageNumber((file_size / PAGE_SIZE) as u32);
+    }
+
+    // performs writes for all provided pages, then waits till the end
+    //
+    // panics if the provided PageNumber is bigger then the max supported one
+    pub fn write(&mut self, pages: Vec<(PageNumber, Box<Page>)>) {
+        let n_write_requests = pages.len();
+        for (pn, page) in pages {
+            let mut command = Some(IoCommand {
+                kind: IoKind::Write(self.store_file.as_raw_fd(), pn.0 as u64, page),
+                handle: self.io_handle_index,
+                user_data: 0,
+            });
+
+            while let Some(c) = command.take() {
+                match self.io_sender.try_send(c) {
+                    Ok(()) => break,
+                    Err(TrySendError::Disconnected(_)) => panic!("I/O leaf store worker dropped"),
+                    Err(TrySendError::Full(c)) => {
+                        command = Some(c);
+                    }
+                }
+            }
+        }
+
+        // wait for completion of all write commands
+        for _ in 0..n_write_requests {
+            let completion = self
+                .io_receiver
+                .recv()
+                .expect("I/O leaf store worker dropped");
+            assert!(completion.result.is_ok());
+        }
+    }
+}
+
+pub struct LeafStoreTx<'a> {
```

Comment:
He-he, can we avoid the `'a` here? 

We really don't want to have it here because we really need to create a tx instance inside `sync`. This happens under mutex. You can't take anything that has a lifetime from under a mutex guard.